lib/factories/default-eventbus.js

```diff
@@ -0,0 +1,373 @@
+ * Any character that does not match [A-Za-z0-9_-] will be replaced with
+ * an underscore.
+ * @param {Object} options
+ * @param {stirng} options.stream_field
+ *      If given, used to extract the event's destination stream name.
+ * @param {string} options.schema_uri_field
+ *      Used to extract the event's schema URI to use as a sanitized stream name.
+ *      This is only used if options.stream_field is not given.
+ */
+function makeExtractStreamName(options) {
+    if (_.isUndefined(options.stream_field)) {
+        const extractSchemaUri = makeExtractSchemaUri(options);
+        const badCharsRegex = new RegExp('[^A-Za-z0-9_-]', 'g');
+        const replacementChar = '_';
+        return (event) => {
+            const schemaUri = extractSchemaUri(event);
+            return schemaUri.replace(badCharsRegex, replacementChar);
+        };
+    } else {
+        return (event) => {
+            return objectGet(event, options.stream_field);
+        };
+    }
+}
+
+/**
+ * Creates a function that extracts a topic (and potentially adds a prefix)
+ * using options.stream_field and options.topic_prefix.
+ * @param {Object} options
+ * @param {stirng} options.stream_field
+ *      Used to extract the event's destination stream name.
+ * @param {stirng} options.topic_prefix
+ *      If given, this will be prefixed to the value extracted from stream_field
+ *      and used as the topic in Kafka.
+ * @return {function(Object): string}
+ */
+function makeExtractTopic(options) {
+    const extractStreamName = makeExtractStreamName(options);
+    return (event) => {
+        const streamName = extractStreamName(event);
+        return options.topic_prefix ?  options.topic_prefix + streamName : streamName;
+    };
+}
+
+/**
+ * Creates a function that extracts a Kafka key, uses options.key_field.
+ * If key_field is not set, this will return undefined instead of a new function.
+ * @param {Object} options
+ * @param {string} options.key_field
+ *      If given, the event's 'key' will be extracted from the event at this dotted object path.
+ * @return {function(Object): *= }
+ */
+function makeExtractKey(options = {}) {
+    if (_.isUndefined(options.key_field)) {
+        return undefined;
+    } else {
+        return event => objectGet(event, options.key_field);
+    }
+}
+
+/**
+ * Creates a function that extracts a kafka partition, useus options.partition_field.
+ * If partition_field is not set, this will return undefined instead of a new function.
+ * @param {Object} options
+ * @param {string} options.partition_field
+ *      If given, the event's 'partition' will be extracted from the event at this
+ *      dotted object path.
+ * @return {function(Object): integer}
+ */
+function makeExtractPartition(options = {}) {
+    if (_.isUndefined(options.partition_field)) {
+        return undefined;
+    } else {
+        return event => _.get(event, options.partition_field);
+    }
+}
```

Comment:
Idem regarding the partitions.

This is really an example of what I quite dislike about including all of this meta information into the event itself. The more metadata we require, the more extra unstandardized fields we have. Now we have `schema_field`, `stream_field`, 'partition_field`. The more features we need the more fields we declare, the tighter the contract with the producers become. 

I do not know a good way out of this, just ranting about how not RESTful this is, but no ideas from me that could support batching..

Reply:
Another idea is that this quite illustrates why I initially wanted to configure this all with objects instead of functions. What if we had `Event` class that would've incapsulated all of these, so we'd end up being like `new Event(payload, config); event.getTopic, event.getPartition, event.toString` etc.. 

Again,do think we need an action here, just ranting

Reply:
HMMMMMM you might be on to something.  I'm going to explore this idea.  One of my aesthetic objections to doing things the REST way in the code is how context has to be carried around in function arguments everywhere, e.g. `validate(event, req, res, config)`, or whatever.  But, what if we did have an abstract Event class that you could extend from?  The Event could carry around the context from the request in non-enumerable properties:

```javascript
class Event {
    constructor(obj, conf, req, res) {
        Object.assign(this, obj);
        Object.defineProperty(this, '_context', { conf, req, res });
    }
    validate() { }
    produce() {}
}
```

Then the default Eventbus could use schema and stream name information from the request context, rather than needing them passed around to the various functions.

Instead of configuring Eventbus with custom validate and produce functions, users could extend the Event class itself, and choose the type of events they expect using something like `event_factory_module`.  IIIIINteresting I'm going to explore this idea.  It might make the using a REST-ier produce easier to implement for the default case, and we can override for whatever batch use case we need to support.

Reply:
Hm, this idea gets tricky though, as I'm still trying to use the Event as the actual event data object with some extra non serializable stuff, instead of as a wrapper class.  What if the incoming event data has fields like '_context', or 'validate'?  I could make some conventions about how non-enumerable fields and methods should be named in Event subclasses, but hm.  Get's tricky.